src/algorithms/rsa.rs

```diff
@@ -154,126 +142,179 @@ pub fn rsa_decrypt_and_check<R: CryptoRngCore + ?Sized>(
+    n: &NonZero<BoxedUint>,
+    e: &BoxedUint,
+    d: &BoxedUint,
+) -> Result<(BoxedUint, BoxedUint)> {
     // Check precondition
-    let two = BigUint::from_u8(2).unwrap();
-    if e <= &two.pow(16u32) || e >= &two.pow(256u32) {
+
+    // Note: because e is at most u64::MAX, it is already
+    // known to be < 2**256
+    if e <= &BoxedUint::from(2u64.pow(16)) {
         return Err(Error::InvalidArguments);
     }
 
     // 1. Let a = (de – 1) × GCD(n – 1, de – 1).
-    let one = BigUint::one();
-    let a = Zeroizing::new((d * e - &one) * (n - &one).gcd(&(d * e - &one)));
+    let bits = d.bits_precision() * 2;
+    let one = BoxedUint::one().widen(bits);
+    let e = e.widen(bits);
+    let d = d.widen(bits);
+    let n = n.as_ref().widen(bits);
+
+    let a1 = &d * &e - &one;
+    let a2 = (&n - &one).gcd(&a1);
+    let a = a1 * a2;
+    let n = n.widen(a.bits_precision());
 
     // 2. Let m = floor(a /n) and r = a – m n, so that a = m n + r and 0 ≤ r < n.
-    let m = Zeroizing::new(&*a / n);
-    let r = Zeroizing::new(&*a - &*m * n);
+    let m = &a / NonZero::new(n.clone()).expect("checked");
+    let r = a - &m * &n;
 
     // 3. Let b = ( (n – r)/(m + 1) ) + 1; if b is not an integer or b^2 ≤ 4n, then output an error indicator,
     //    and exit without further processing.
-    let modulus_check = Zeroizing::new((n - &*r) % (&*m + &one));
-    if !modulus_check.is_zero() {
+    let modulus_check = (&n - &r) % NonZero::new(&m + &one).expect("adding 1");
+    if (!modulus_check.is_zero()).into() {
         return Err(Error::InvalidArguments);
     }
-    let b = Zeroizing::new((n - &*r) / (&*m + &one) + one);
+    let b = ((&n - &r) / NonZero::new(&m + &one).expect("adding one")) + one;
 
-    let four = BigUint::from_u8(4).unwrap();
-    let four_n = Zeroizing::new(n * four);
-    let b_squared = Zeroizing::new(b.pow(2u32));
-    if *b_squared <= *four_n {
+    let four = BoxedUint::from(4u32);
+    let four_n = &n * four;
+    let b_squared = b.square();
+
+    if b_squared <= four_n {
         return Err(Error::InvalidArguments);
     }
-    let b_squared_minus_four_n = Zeroizing::new(&*b_squared - &*four_n);
+    let b_squared_minus_four_n = b_squared - four_n;
 
     // 4. Let ϒ be the positive square root of b^2 – 4n; if ϒ is not an integer,
     //    then output an error indicator, and exit without further processing.
-    let y = Zeroizing::new(sqrt((*b_squared_minus_four_n).clone()));
+    let y = b_squared_minus_four_n.sqrt();
 
-    let y_squared = Zeroizing::new(y.pow(2u32));
+    let y_squared = y.square();
     let sqrt_is_whole_number = y_squared == b_squared_minus_four_n;
     if !sqrt_is_whole_number {
         return Err(Error::InvalidArguments);
     }
-    let p = (&*b + &*y) / &two;
-    let q = (&*b - &*y) / two;
+
+    let bits = core::cmp::max(b.bits_precision(), y.bits_precision());
+    let two = NonZero::new(BoxedUint::from(2u64))
+        .expect("2 is non zero")
+        .widen(bits);
```

Comment:
Something which makes this pattern better would be nice. We have various `NonZero::new_unwrap` constructors which allow you to infallibly construct a `NonZero`.

Reply:
```
278 |     let two = NonZero::new_unwrap(2).widen(bits);
    |                        ^^^^^^^^^^ multiple `new_unwrap` found
    |
    = note: candidate #1 is defined in an impl for the type `crypto_bigint::NonZero<Limb>`
    = note: candidate #2 is defined in an impl for the type `crypto_bigint::NonZero<crypto_bigint::Uint<LIMBS>>`
```

seems to not exist for the`BoxedUint` version

Reply:
@dignifiedquire aah, yeah the idea would be to const construct a `NonZero<Uint>` first and either convert it to a `NonZero<BoxedUint>` first or make it possible to use `NonZero<Uint>` as an operand.